### ISSUE TITLE
Non exact-zero value for fake quantizer ops

### ISSUE THREAD

**Issue by CNugteren** [tensorflow/tensorflow#47193]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.4.0-49-g85c8b2a817f 2.4.1
- Python version: 3.8.5
- CUDA/cuDNN version: 11.0 / 8.1.0
- GPU model and memory: Tesla T4 and GeForce GTX 1650

**Describe the current behavior**

The [tf.quantization.fake_quant_with_min_max_vars](https://www.tensorflow.org/api_docs/python/tf/quantization/fake_quant_with_min_max_vars) (and similar) does not include an exact zero result. Depending on the `min` and `max` parameters these values are actually slightly above zero or below zero, e.g. -0.0000002 or 0.0000006.

**Describe the expected behavior**

An exact zero should be returned and it should not be dependent on the `min` and `max` parameters. Or at least, this was the behaviour in pre-TF 2.4.0 tensorflow versions on the CPU. It is also mentioned in [this article in section 2.1](https://arxiv.org/abs/1712.05877).

**Standalone code to reproduce the issue**

The simplest snippet to reproduce the issue is:
```python
print(tf.quantization.fake_quant_with_min_max_vars(0.0, -25, 20, num_bits=8))
```
with TF 2.4.1 on (CPU or GPU) I get:
```python
tf.Tensor(-3.874302e-07, shape=(), dtype=float32)
```
with TF 2.3.2 on the CPU I get:
```python
tf.Tensor(0.0, shape=(), dtype=float32)
```
and with TF 2.3.2 on the GPU I get:
```python
tf.Tensor(-3.874302e-07, shape=(), dtype=float32)
```

A more realistic example would look something like this:
```python
data = tf.random.normal(shape=(4096,), mean=0.0, stddev=1.0)
result = tf.quantization.fake_quant_with_min_max_vars(data, -21, 23, num_bits=8).numpy()
print([result[index] for index in range(result.shape[0]) if -1e-4 < result[index] < 1e-4][0])
```

Colab link: https://colab.research.google.com/drive/1swMaNOZpQp1psllRguWH13YHYOpf2Kqa?usp=sharing

**Other info / logs**

I've tested this in various settings on various machines, and it seems to work in certain scenarios:

| platform | TF 2.2.x | TF 2.3.x | TF 2.4.x |
|------------|-------------|------------|-------------|
| Mac      | ✅  | ✅  | ✅ |
| Ubuntu 20.04 CPU | ✅ | ✅ | ❌ |
| Ubuntu 20.04 GPU | ❌ | ❌ | ❌ | 

Here ❌ means it is not exact zero, and ✅ means it is exact zero.

It also seems to work if the function is enclosed in a `tf.function` and XLA is enabled by passing in `experimental_compile=True`.

I could not find any recent commits to the relevant files:
* https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fake_quant_ops.cc
* https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fake_quant_ops_gpu.cu.cc
* https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fake_quant_ops_functor.h

So, most likely the issue is in one of the TF utility functions or in an external library (e.g. Eigen) or in the way the TensorFlow wheels are compiled.

**Comment by ravikyram** [comment]
Was able to reproduce the issue with  TF v2.4. In TF-nightly using cpu i get output (`<tf.Tensor: shape=(), dtype=float32, numpy=0.0>`) .Please find the gist of it [here](https://colab.research.google.com/gist/ravikyram/5ca92301c765d1a163d2b1efacde0a92/untitled673.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
@CNugteren Agree. This was an issue with `TF2.4` but I think this was resolved in recent `tf-nightly`. I cannot reproduce the error with `tf-nightly` and `tf-nightly-gpu`. 

Please verify with `tf-nightly` and let us know how it progresses. Thanks!

**Comment by lgeiger** [comment]
@jvishnuvardhan Thanks for taking a look.
I checked out your GPU notebook but it looks like the computation actually doesn't run on the GPU but on the CPU instead even though a GPU is connected to the runtime. I am not sure whether this is an issue with Colab or the latest nightly, but we'll still need to check whether this issue is also fixed in the GPU kernel.
Please see [this gist](https://colab.research.google.com/drive/1l3VCKACv81rDM_B88czUDbWftid8Q027?usp=sharing) for reference.

**Comment by CNugteren** [comment]
Thanks for looking into this. On my own local machine it does seem to use the GPU properly, but there the value is still not exact zero. It does work on the CPU now indeed, but not on the GPU:
```python
$ cat test_quant.py
import tensorflow as tf
print(tf.__version__)
with tf.device("GPU"):
    print(tf.quantization.fake_quant_with_min_max_vars(0.0, -20, 20, num_bits=8).device)
    print(tf.quantization.fake_quant_with_min_max_vars(0.0, -20, 20, num_bits=8))
```
```bash
$ export CUDA_VISIBLE_DEVICES= && python test_quant.py 
(...)
2.5.0-dev20210218
(...)
2021-02-19 10:46:15.910727: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
(...)
/job:localhost/replica:0/task:0/device:CPU:0
tf.Tensor(0.0, shape=(), dtype=float32)
```
```bash
$ export CUDA_VISIBLE_DEVICES=0 && python test_quant.py 
(...)
2.5.0-dev20210218
(...)
2021-02-19 10:46:12.245787: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1456] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2208 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1650, pci bus id: 0000:01:00.0, compute capability: 7.5)
/job:localhost/replica:0/task:0/device:GPU:0
tf.Tensor(-4.917383e-07, shape=(), dtype=float32)
```

**Comment by teijeong** [comment]
Hello, sorry for late reply. Seems like it's been resolved in the `tf-nightly`. Can you check again?

**Comment by CNugteren** [comment]
> Hello, sorry for late reply. Seems like it's been resolved in the `tf-nightly`. Can you check again?

It does not seem to be resolved, at least not on my specific GPU/CUDA combination:

```python
$ cat test_quant.py
import tensorflow as tf
print(tf.__version__)
with tf.device("GPU"):
    print(tf.quantization.fake_quant_with_min_max_vars(0.0, -20, 20, num_bits=8).device)
    print(tf.quantization.fake_quant_with_min_max_vars(0.0, -20, 20, num_bits=8))
```

```bash
$ export CUDA_VISIBLE_DEVICES= && python test_quant.py
(...)
2.5.0-dev20210323
(...)
2021-03-23 10:50:12.857546: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
(...)
/job:localhost/replica:0/task:0/device:CPU:0
tf.Tensor(0.0, shape=(), dtype=float32)
```
```bash
$ export CUDA_VISIBLE_DEVICES=0 && python test_quant.py
(...)
2.5.0-dev20210323
(...)
2021-03-23 10:55:37.606288: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce GTX 1650 computeCapability: 7.5
coreClock: 1.56GHz coreCount: 16 deviceMemorySize: 3.82GiB deviceMemoryBandwidth: 119.24GiB/s
(...)
2021-03-23 10:55:37.983165: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 2131 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1650, pci bus id: 0000:01:00.0, compute capability: 7.5)
/job:localhost/replica:0/task:0/device:GPU:0
tf.Tensor(-4.917383e-07, shape=(), dtype=float32)

```

**Comment by Xhark** [comment]
Logic on fake quant try to make zero to be one of the quantized integer, so result should be close to zero.
But due to internal logic only saves nudged_min (map to min value) and scale, It can have small numerical errors.

I think it's not possible to have exact zero on GPU because usually CUDA ops are not deterministic. 
(has random small errors)

We don't prioritize this kinds of small numeric error.
(may don't have any way to fix it on GPU. it requires deterministic like this : https://github.com/tensorflow/community/blob/master/rfcs/20210119-determinism.md)

Do you have any task that this bug is critical?

**Comment by lgeiger** [comment]
@Xhark thanks for the answer. I agree that numerical stability is not a priority for floating point neural networks. However, I think for the ops that emulate quantized networks the story can be a bit different.

> Do you have any task that this bug is critical?

We work on Binarized Neural Networks which we train using [larq/larq](https://github.com/larq/larq) and deploy using TFLite and [larq/compute-engine](https://github.com/larq/compute-engine) on edge devices. In these networks weights and activations tend to be binary values `{-1, +1}`. A common pattern is the use of higher precision int8 or floating point shortcuts that will be binarized using a `sign` function before the next binary convolution, such that strictly negative values are mapped to `-1` and the rest to `+1`.
Unfortunately this causes issues in cases where int8 quantization doesn't correctly represent `0`, since huge quantization errors will be introduced in the following binarization operation. Values might flip from `+1` to `-1` due to these numerical instabilities during int8 quantization which can completely destroy the accuracy of these models when converted to TFLite.

While our use case is particular to low bit networks, the issue can also be problematic for general `int8` quantized networks:
Both https://arxiv.org/pdf/1712.05877.pdf (section 2.1) and https://arxiv.org/pdf/2004.09602.pdf (section 3.1.1) mention that it is important for the int8 quantization scheme to represent zero exactly and https://arxiv.org/pdf/1806.08342.pdf (section 2.1) explicitly states:
>  Zero-point is an integer, ensuring that zero is quantized with no error. This is important to ensure that common operations like zero padding do not cause quantization error.

If the ops used in quantization aware training do not satisfy this constrain the error can accumulate over many layers, potentially leading to a bias in the predictions.

Currently this doesn't seem to be an issue when using [`tf.quantization.quantize_and_dequantize`](https://www.tensorflow.org/api_docs/python/tf/quantization/quantize_and_dequantize).
Is this op preferred when doing quantization aware training in TensorFlow?

Most docs point to `tf.quantization.fake_quant_with_min_max_vars` which is also used in TF-MOT but also has limited support for mixed precising float16 training and suffers from the numerical issues discussed here.
While `tf.quantization.quantize_and_dequantize` seems to work fine, it is deprecated in favour of [`tf.quantization.quantize_and_dequantize_v2`](https://www.tensorflow.org/api_docs/python/tf/quantization/quantize_and_dequantize_v2). `tf.quantization.quantize_and_dequantize_v2` in turn is not yet supported by the TFLite converter. I opened #47225 some time ago which is a requirement for adding support for `tf.quantization.quantize_and_dequantize_v2` to the TFLite converter in a followup PR, but it seems to be stuck in the pipeline.

@Xhark I am curious to hear your thoughts on this and would also appreciate some pointers on the difference between the mentioned `tf.quantization` ops.

**Comment by lgeiger** [comment]
Just for reference it looks like a similar issue has been reported in https://github.com/tensorflow/model-optimization/issues/635

And it seems like #48580 might fix this issue, but the PR is still waiting for review.

**Comment by Xhark** [comment]
@lgeiger 
Thanks for your comment. It might be better to add epsilon value for your binarizing use case if it possible.
We assume that zero-point is representable with integer number, and all optimization / implementation also assume that we have integer zero-point. (e.g. zero-padding is easily possible on quantized domain, some pre-compute optimizations for the symmetric quantized weight assume we have integer zero-point.)

I agree that it can causes accumulate floating numeric errors, but it also happens for other than zero value. (Of course, zero might be more critical than other values for some use cases.)

 - tf.quantization
 For QAT, all TF level training & inference only uses FakeQuant. There's no actual quantization yet. It only actually quantized when you convert the model to TFLite format. I think tf.quantization.quantize_and_dequantize_v2 is only for backward compatibility. But very interesting that it's better than FakeQuant in terms of numerical stability. Theoretically, FakeQuant also has same level of the numerical stability. It seems need to do more research on it.

 - About pull request
I think that pull request that you linked will solve more cases, but we can't guarantee it works for all kinds of accelerator. (some compiler optimizations also can changes numerical stability.) It potentially can causes a big problem if your code is very sensitive for this kinds of near-zero numeric error, so the best way is makes it less error sensitive if possible.

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue with TF v2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/6b04ad9466aa0beb140834d9df5e3051/untitled125.ipynb?authuser=1)..Thanks !

**Comment by CNugteren** [comment]
Yes this issue can be closed. I cross-checked with my colleague on macOS that it works as expected there as well.
